File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/BaseEditTimetableFragment.java
package com.happs.ximand.ringcontrol.view.fragment;

import androidx.annotation.NonNull;
import androidx.databinding.ViewDataBinding;

import com.happs.ximand.ringcontrol.R;
import com.happs.ximand.ringcontrol.model.object.Lesson;
import com.happs.ximand.ringcontrol.view.adapter.EditTimetableRecyclerViewAdapter;
import com.happs.ximand.ringcontrol.viewmodel.fragment.BaseEditTimetableViewModel;

import java.util.List;

public abstract class BaseEditTimetableFragment<VM extends BaseEditTimetableViewModel,
        D extends ViewDataBinding>
        extends BaseFragmentWithRecyclerView<VM, D, Lesson, EditTimetableRecyclerViewAdapter> {

    public BaseEditTimetableFragment(int layoutId) {
        super(layoutId, R.menu.menu_toolbar_editing);
    }

    @Override
    protected void onPreViewModelAttaching(@NonNull VM viewModel) {
        viewModel.getLessonsLiveData().observe(
                getViewLifecycleOwner(), this::initAdapter
        );
        viewModel.getDetailEditingLiveData().observe(
                getViewLifecycleOwner(), this::updateDetailEditingStatus
        );
        viewModel.setCorrectnessCheck(this::isAdapterDataCorrect);
    }

    @Override
    protected EditTimetableRecyclerViewAdapter createNewAdapter(List<Lesson> items) {
        return new EditTimetableRecyclerViewAdapter(items);
    }

    @Override
    protected void onPreAttachRecyclerViewAdapter(EditTimetableRecyclerViewAdapter adapter) {
        getViewModel().getAddLessonEvent().observe(getViewLifecycleOwner(),
                aVoid -> adapter.addEmptyLesson()
        );
        getViewModel().getRemoveLessonEvent().observe(getViewLifecycleOwner(),
                aVoid -> adapter.removeLastLesson()
        );
    }

    private void updateDetailEditingStatus(boolean status) {
        EditTimetableRecyclerViewAdapter adapter = getAdapter();
        if (adapter != null) {
            adapter.setDetailEditingStatus(status);
        }
    }

    private boolean isAdapterDataCorrect() {
        EditTimetableRecyclerViewAdapter adapter = getAdapter();
        if (adapter != null) {
            return adapter.isAllLinesCorrect();
        }
        return false;
    }

}

File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/BaseEditTimetableFragment.kt
package com.happs.ximand.ringcontrol.view.fragment

import androidx.databinding.ViewDataBinding
import androidx.lifecycle.Observer
import com.happs.ximand.ringcontrol.R
import com.happs.ximand.ringcontrol.databinding.ItemEditLessonBinding
import com.happs.ximand.ringcontrol.model.`object`.timetable.Lesson
import com.happs.ximand.ringcontrol.view.adapter.EditTimetableRecyclerViewAdapter
import com.happs.ximand.ringcontrol.viewmodel.fragment.BaseEditTimetableViewModel

abstract class BaseEditTimetableFragment<VM : BaseEditTimetableViewModel, D : ViewDataBinding>(layoutId: Int) : BaseFragmentWithRecyclerView<VM, D, Lesson, ItemEditLessonBinding, EditTimetableRecyclerViewAdapter>(layoutId, R.menu.menu_toolbar_editing) {

    override fun onPreViewModelAttaching(viewModel: VM) {
        viewModel.lessonsLiveData.observe(viewLifecycleOwner,
                Observer { initAdapter(it) })
        viewModel.detailEditingLiveData.observe(viewLifecycleOwner,
                Observer { updateDetailEditingStatus(it) })
        viewModel.correctnessCheck = { isAdapterDataCorrect() }
    }

    override fun createNewAdapter(items: MutableList<Lesson>): EditTimetableRecyclerViewAdapter {
        return EditTimetableRecyclerViewAdapter(items)
    }

    override fun onPreAttachRecyclerViewAdapter(adapter: EditTimetableRecyclerViewAdapter) {
        requireViewModel().addLessonEvent.observe(viewLifecycleOwner,
                Observer { adapter.addEmptyLesson() }
        )
        requireViewModel().removeLessonEvent.observe(viewLifecycleOwner,
                Observer { adapter.removeLastLesson() }
        )
    }

    private fun updateDetailEditingStatus(status: Boolean) {
        val adapter = adapter
        adapter?.setDetailEditingStatus(status)
    }

    private fun isAdapterDataCorrect(): Boolean {
        val adapter = adapter
        return adapter?.isAllLinesCorrect() ?: false
    }
}